### DISCUSSION TITLE
Getting email_verified From Auth0

### DISCUSSION THREAD

**Discussion by github0013** [nextauthjs/next-auth#2235] [Help] [answered]
## Question 💬

I am working on implementing Auth0 provider (using only "database connection" that comes by default).

https://next-auth.js.org/providers/auth0 

> The Auth0 Provider comes with a set of default options:
> - Auth0 Provider options
>
> You can override any of the options to suit your own use case.

And this is the defaults.
https://github.com/nextauthjs/next-auth/blob/718f2537cbde29ea39ef0a734943ac0f11e009be/src/providers/auth0.js

### Auth0 user Profile
https://auth0.com/docs/api/authentication#user-profile

> The email scope value requests access to the email and email_verified Claims.

It seems, if I edit the profile here

https://github.com/nextauthjs/next-auth/blob/718f2537cbde29ea39ef0a734943ac0f11e009be/src/providers/auth0.js#L12

I could get `email_verified` attribute, but I don't know how I should go about changing the profile's return object.

```ts
export default NextAuth({
  providers: [
    Providers.Auth0({
      profile(profile: any) {
        return {
          id: profile.sub,
          name: profile.nickname,
          email: profile.email,
          email_verified: profile.email_verified,
          image: profile.picture,
        }
      },
```

Tried like this, but I didn't get `session.user.email_verified` attribute.

How does it work?

**Comment by balazsorban44** [comment, marked as answer]
Check https://next-auth.js.org/configuration/callbacks#session-callback as well

**Reply by github0013** [reply]
https://github.com/nextauthjs/next-auth/discussions/2226#discussioncomment-903324

This really helped me to understand how to code.

```ts
export default NextAuth({
  callbacks: {
    async jwt(token, user, _account, _profile, _isNewUser) {
      if (user) {
        token.email_verified = user.email_verified // pull that out
      }
      return token
    },

    async session(session, token) {
      session.auth0 = {
        email_verified: Boolean(token.email_verified), // and put it in
      }
      return session
    },
  },
  providers: [
    Providers.Auth0({
      profile(profile: any, tokens) {
        return {
          id: profile.sub,
          name: profile.nickname,
          email: profile.email,
          image: profile.picture,
          email_verified: profile.email_verified, // add the attribute here
        }
      },
      clientId: process.env.AUTH0_CLIENT_ID,
      clientSecret: process.env.AUTH0_CLIENT_SECRET,
      domain: process.env.AUTH0_DOMAIN,
    }),
  ],
```

<img width="455" alt="My_Dashboard" src="https://user-images.githubusercontent.com/809378/123188704-71f6f800-d4d7-11eb-9447-53d222b40faa.png">


https://github.com/nextauthjs/next-auth/discussions/2226#discussioncomment-904672

I also feel there should be some kind of documentation around this session and JWT.
I think the majority of the `next-auth` users are not familiar with JWT verification processes (that is why we are using a helpful library).

If there were a diagram showing how JWT is provided from the OAuth provider, and then how `next-auth` is using it to create session object, it would help many to just get in coding without asking.

**Reply by balazsorban44** [reply]
We do document the callbacks here https://next-auth.js.org/configuration/callbacks

If you think the documentation could be improved, please don't hesitate to open a PR with your suggestions! 🙏

**Reply by github0013** [reply]
Yes, it does mention about how to add extra attributes into the session object.  However, it seems it's not that obvious how it should be done.

https://next-auth.js.org/getting-started/client#custom-client-session-handling
If you can add an example like you did above, I think we'll have less of the same questions over this.

**Comment by juicemia** [comment]
Is there any updated guidance on this? I tried the above code but it doesn't work. The `session` object doesn't have an `auth0` property to set so Typescript compiling fails. The `user` parameter on the `jwt` function also doesn't have an `email_verified` parameter.

EDIT: I'm dropping all my code in here:

```
export const authOptions: NextAuthOptions = {
  providers: [
    Auth0Provider({
      clientId: process.env.AUTH0_CLIENT_ID,
      clientSecret: process.env.AUTH0_CLIENT_SECRET,
      issuer: process.env.AUTH0_ISSUER,
      profile(profile: any, tokens) {
        console.log("profile")
        console.log(profile);
        console.log(tokens);

        return {
          id: profile.sub,
          name: profile.nickname,
          email: profile.email,
          image: profile.picture,
          email_verified: profile.email_verified,
        }
      },
    })
  ],
  adapter: MongoDBAdapter(mongoStore.init(), {
    databaseName: NEXTAUTH_MONGODB_DATABASE
  }),
  callbacks: {
    async session({ session, token }) {
      console.log("session")
      console.log(JSON.stringify(session))
      console.log(JSON.stringify(token))

      return session
    },
  },
}
```

This is what I ended up with. `emailVerified` is still `null` in MongoDB for a user that's verified. `null` is suspicious to me I would have expected it to be false. Almost makes me think that the provider and adapter aren't hooked up correctly.
